Add Spoofed Container Runtime

HonakerM opened the pull request:
<!--  Thanks for sending a pull request!

Please be aware that we're following the Kubernetes guidelines of contributing
to this project. This means that we have to use this mandatory template for all
of our pull requests.

Please also make sure you've read and understood our contributing guidelines
(https://github.com/cri-o/cri-o/blob/main/CONTRIBUTING.md) as well as ensuring
that all your commits are signed with `git commit -s`.

Here are some additional tips for you:

- If this is your first time, please read our contributor guidelines:
  https://git.k8s.io/community/contributors/guide#your-first-contribution and
  developer guide
  https://git.k8s.io/community/contributors/devel/development.md#development-guide
- Please label this pull request according to what type of issue you are
  addressing, especially if this is a release targeted pull request. For
  reference on required PR/issue labels, read here:
  https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
- If you want *faster* PR reviews, read how:
  https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
- If the PR is unfinished, see how to mark it:
  https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Uncomment only one `/kind <>` line, hit enter to put that in a new line, and
remove leading whitespace from that line:
-->

<!--
/kind api-change
/kind bug
/kind ci
/kind cleanup
/kind dependency-change
/kind deprecation
/kind design
/kind documentation
/kind failing-test
/kind feature
/kind flake
/kind other
-->
/kind feature

#### What this PR does / why we need it:
This PR adds support for a new spoofed CRIO runtime which can mock the creation of containers and sandboxes. These containers report status as usual and can be interacted with using crictl; however, they do not start runc containers saving resources and image pull time. The original idea behind this functionality was to speed up development and testing of [operators](https://kubernetes.io/docs/concepts/extend-kubernetes/operator/), which require orchestrating a large number of pods that can have lengthy startup and image pull times. 

#### Which issue(s) this PR fixes:

<!--
Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
-->

<!--
Fixes #
or
None
-->

#### Special notes for your reviewer:

This is my first PR for CRIO so any feedback would be greatly appreciated. I opened this PR after [some discussion on slack](https://kubernetes.slack.com/archives/CAZH62UR1/p1668093677867629?thread_ts=1668031829.089849&cid=CAZH62UR1) to get more feedback on the functionality.

#### Does this PR introduce a user-facing change?

<!--
If no, just write `None` in the release-note block below. If yes, a release note
is required: Enter your extended release note in the block below. If the PR
requires additional action from users switching to the new release, include the
string "action required".

For more information on release notes see:
https://git.k8s.io/community/contributors/guide/release-notes.md
-->
no user facing change but it does add more configuration options. I'm not sure if that counts as a user-facing change.
 
```release-note
None
```

HonakerM commented on pkg/config/config.go:
```diff
@@ -233,6 +234,13 @@ type RuntimeConfig struct {
 	// DropInfraCtr determines whether the infra container is dropped when appropriate.
 	DropInfraCtr bool `toml:"drop_infra_ctr"`
 
+	//TODO Beter place for these?
```
I'm not very familiar with CRIO's config format so please suggest better names and/or location for these values.

haircommander commented:
Thanks for the PR @HonakerM 

This feels like a natural fit for a new runtime class. a new runtime handler could be created in `internal/oci/`, and instead of a top-level config field (Spoofed and SpoofedPassThrough), the admin could just specify `default_runtime = "spoofed"`

HonakerM commented:
@haircommander ohhhhhh! That makes alot more sense! Just to confirm you're talking about adding a new file like `internal/oci/runtime_spoofed.go` which just focuses on spoofing containers. 

And on a different note do you have any suggestion regarding the `SpoofedPassThrough` configuration? From my understanding if the admin selects `default_runtime = "spoofed"` all containers would be spoofed regardless.

haircommander commented:
> @haircommander ohhhhhh! That makes alot more sense! Just to confirm you're talking about adding a new file like `internal/oci/runtime_spoofed.go` which just focuses on spoofing containers.

Yeah exactly 

> And on a different note do you have any suggestion regarding the `SpoofedPassThrough` configuration? From my understanding if the admin selects `default_runtime = "spoofed"` all containers would be spoofed regardless. Are there configuration options for specific runtime classes?

a pod can opt out of it by setting a different runtime through kubernetes runtime classes

HonakerM commented:
@haircommander I've updated my PR to implement a separate `spoofed` runtime. Please let me know what you think. I'm not sure where to add test cases for it.... I'm not to familiar with go's test suite just yet so any tips would be greatly appreciated.

haircommander commented on internal/oci/runtime_spoofed.go:
```diff
@@ -0,0 +1,245 @@
+func (r *runtimeSpoofed) CreateContainer(ctx context.Context, c *Container, cgroupParent string, restore bool) (retErr error) {
+
+	// Start sleep command to create process space for container
+	cmd := cmdrunner.Command("/bin/sleep","infinity") // nolint: gosec
```
I would like to figure out a way to avoid having this process. IT doesn't looks like it's being stopped anywhere, and adds additional overhead. What was the motivation for having it?

HonakerM replied:
The original motivation was to create a process space that the CNI could still hook into as I started running into issues if the PID of the container was 0. However, I'm not sure if it's needed as we could just use the PID of crio.

If this duplicate process is actually needed I can add in process termination/cleanup into` StopContainer`.

haircommander commented:
/ok-to-test

HonakerM commented:
@haircommander thank you for all of your help, I'm sorry for my multiple questions. I'm a little confused on how to add more runtime tests. I see that there is already a set of container tests in `test/ctr.bats` and to run them using the `spoofed` runtime I can use `make RUNTIME_TYPE=spoofed localintegration`. Other than updating the tests to skip when using spoofed [similarly to skipping the pod runtime](https://github.com/cri-o/cri-o/blob/7591cffa2a8b3bfc77190a45db54d90d4e45f61d/test/ctr.bats#L186) is there anything else I need to do?

HonakerM commented:
Sorry about that! I'm still working on adding tests but am busy with end of year madness. I'll hopefully have this PR in a reviewable state in January

HonakerM commented:
/remove-lifecycle stale

@haircommander sorry for the delay I've finally got back around to this PR. I've added tests and added process cleanup. Any suggestions regarding [the issue regarding process namespace](https://github.com/cri-o/cri-o/pull/6375#discussion_r1021866210) would be greatly appreciated.
